File: Linked-List/LinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self,value):
        new_node = Node(value)
        self.head = new_node
        self.tail = new_node
        self.length = 1

    def print_list(self):
        temp = self.head
        while temp is not None:
            print(temp.value)
            temp = temp.next
    
    def append(self, value):
        new_node = Node(value)
        if self.head is  None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1
        return True
    
    def pop(self):
        if self.head is None:
            return None
        elif self.head.next is None:
            popped_value = self.head.value
            self.head = None
            self.tail = None
            self.length -= 1
            return popped_value
        else:
            current = self.head
            while current is not None:
                if current.next is None:
                    self.tail = last
                    self.tail.next = None
                    self.length -= 1
                    return current.value
                last = current
                current = current.next
    
    def pop(self):
        if self.length == 0:
            return None
        temp = self.head
        prev = self.head
        while temp.next:
            prev = temp
            temp = temp.next
        self.tail = prev
        self.tail.next = None
        self.length -= 1
        if self.length == 0:
            self.head = None
            self.tail = None
        return temp
        
    def prepend(self, value):
        new_node = Node(value)
        if self.length == 0:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head = new_node
        self.length += 1
        return True
    
    def pop_first(self):
        if self.length == 0:
            return None
        temp = self.head
        self.head = self.head.next #also make head None if only one node exists
        temp.next = None
        self.length -= 1
        if self.length == 0:
            self.tail = None
        return temp
    
    def get(self,index):
        if index < 0 or index >= self.length:
            return None
        temp = self.head
        for _ in range(index):
            temp = temp.next
        return temp
    
    def set_value(self,index,value):
        temp = self.get(index)
        if temp:
            temp.value = value
            return True
        return False

    def insert(self,index,value):
        if index < 0 or index > self.length:
            return False
        if index == 0:
            return self.prepend(value)
        if index == self.length:
            return self.append(value)
        new_node = Node(value)
        prev = self.get(index - 1)
        new_node.next = prev.next
        prev.next = new_node
        self.length += 1
        return True
    
    def remove(self, index):
        if index < 0 or index >= self.length:
            return None
        if index == 0:
            return self.pop_first()
        if index == self.length - 1:
            return self.pop()
        prev = self.get(index - 1)
        removed_node  = prev.next
        prev.next = removed_node.next
        removed_node.next = None
        self.length -=1
        return removed_node
    
    def reverse(self):
        if self.head is None:
            return
        temp = self.head
        self.head = self.tail
        self.tail = temp
        after = temp.next
        before = None
        for _ in range(self.length):
            after = temp.next
            temp.next = before
            before = temp
            temp = after 
    
my_linked_list = LinkedList(2)

# my_linked_list.append(2)
# my_linked_list.append(3)
# # print(my_linked_list.tail.next)
# my_linked_list.print_list()
# print()
# my_linked_list.pop()
# my_linked_list.pop()
# my_linked_list.pop()
# my_linked_list.pop()
# my_linked_list.pop()
# # my_linked_list.pop()
# my_linked_list.print_list()
# print(my_linked_list.length)
my_linked_list.prepend(1)
my_linked_list.print_list()
print(my_linked_list.length)

File: Linked-List/LinkedList.java
class Node {
    int value;
    Node next;
    
    public Node(int value) {
        this.value = value;
        this.next = null;
    }
}

public class LinkedList {
    private Node head;
    private Node tail;
    private int length;
    
    public LinkedList(int value) {
        Node newNode = new Node(value);
        this.head = newNode;
        this.tail = newNode;
        this.length = 1;
    }
    
    public void printList() {
        Node temp = this.head;
        while (temp != null) {
            System.out.println(temp.value);
            temp = temp.next;
        }
    }
    
    public boolean append(int value) {
        Node newNode = new Node(value);
        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
        return true;
    }
    
    public Node pop() {
        if (this.length == 0) {
            return null;
        }
        Node temp = this.head;
        Node prev = this.head;
        while (temp.next != null) {
            prev = temp;
            temp = temp.next;
        }
        this.tail = prev;
        this.tail.next = null;
        this.length--;
        if (this.length == 0) {
            this.head = null;
            this.tail = null;
        }
        return temp;
    }
    
    public boolean prepend(int value) {
        Node newNode = new Node(value);
        if (this.length == 0) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        return true;
    }
    
    public Node popFirst() {
        if (this.length == 0) {
            return null;
        }
        Node temp = this.head;
        this.head = this.head.next;
        temp.next = null;
        this.length--;
        if (this.length == 0) {
            this.tail = null;
        }
        return temp;
    }
    
    public Node get(int index) {
        if (index < 0 || index >= this.length) {
            return null;
        }
        Node temp = this.head;
        for (int i = 0; i < index; i++) {
            temp = temp.next;
        }
        return temp;
    }
    
    public boolean setValue(int index, int value) {
        Node temp = this.get(index);
        if (temp != null) {
            temp.value = value;
            return true;
        }
        return false;
    }
    
    public boolean insert(int index, int value) {
        if (index < 0 || index > this.length) {
            return false;
        }
        if (index == 0) {
            return this.prepend(value);
        }
        if (index == this.length) {
            return this.append(value);
        }
        Node newNode = new Node(value);
        Node prev = this.get(index - 1);
        newNode.next = prev.next;
        prev.next = newNode;
        this.length++;
        return true;
    }
    
    public Node remove(int index) {
        if (index < 0 || index >= this.length) {
            return null;
        }
        if (index == 0) {
            return this.popFirst();
        }
        if (index == this.length - 1) {
            return this.pop();
        }
        Node prev = this.get(index - 1);
        Node removedNode = prev.next;
        prev.next = removedNode.next;
        removedNode.next = null;
        this.length--;
        return removedNode;
    }
    
    public void reverse() {
        Node temp = this.head;
        this.head = this.tail;
        this.tail = temp;
        Node after = temp.next;
        Node before = null;
        for (int i = 0; i < this.length; i++) {
            after = temp.next;
            temp.next = before;
            before = temp;
            temp = after;
        }
    }
    
    // Getter methods for testing
    public int getLength() {
        return this.length;
    }
    
    public Node getHead() {
        return this.head;
    }
    
    public Node getTail() {
        return this.tail;
    }
    
    // Main method for testing
    public static void main(String[] args) {
        LinkedList myLinkedList = new LinkedList(2);
        
        myLinkedList.prepend(1);
        myLinkedList.printList();
        System.out.println("Length: " + myLinkedList.getLength());
    }
}
